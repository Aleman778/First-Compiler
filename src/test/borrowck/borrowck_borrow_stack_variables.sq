// ignore

// Testing the borrowing rules

fn borrow_same_mut_twice() {
    let mut x: i32 = 10;
    let a: &mut i32 = &mut x;
    let _b: &mut i32 = &mut x;
    *a;
}


fn borrow_same_mut_imm() {
    let mut x: i32 = 10;
    let a: &mut i32 = &mut x;
    let _b: &i32 = &x;
    *a;
}


fn borrow_same_imm_mut() {
    let mut x: i32 = 10;
    let a: &i32 = &x;
    let _b: &mut i32 = &mut x;
    *a;
}


fn borrow_same_imm_imm() {
    let mut x: i32 = 10;
    let a: &i32 = &x;
    let _b: &i32 = &x;
    *a;
}


fn main() {}
