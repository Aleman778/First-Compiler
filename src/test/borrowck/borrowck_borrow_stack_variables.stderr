error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> $DIR/borrowck_borrow_stack_variables.sq:8:24
   |
LL |     let a: &mut i32 = &mut x;
   |                       ------ first mutable borrow occurs here
LL |     let _b: &mut i32 = &mut x;
   |                        ^^^^^^ second mutable borrow occurs here
LL |     *a;
   |     -- first borrow later used here

error[E0502]: cannot borrow `x` as immutable because it is also borrowed as mutable
  --> src/test/borrowck/borrowck_borrow_stack_variables.sq:16:20
   |
LL |     let a: &mut i32 = &mut x;
   |                       ------ mutable borrow occurs here
LL |     let _b: &i32 = &x;
   |                    ^^ immutable borrow occurs here
LL |     *a;
   |     -- mutable borrow later used here

error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
  --> src/test/borrowck/borrowck_borrow_stack_variables.sq:24:24
   |
LL |     let a: &i32 = &x;
   |                   -- immutable borrow occurs here
LL |     let _b: &mut i32 = &mut x;
   |                        ^^^^^^ mutable borrow occurs here
LL |     *a;
   |     -- immutable borrow later used here